### ISSUE TITLE
VB.NET: The compiler should report the use of wrong type characters on variables (like it did before)

### ISSUE THREAD

**Issue by slepmog** [dotnet/roslyn#44724]
**Version Used**: 
VS 15.9.23, VS 16.5

**Steps to Reproduce**:

Compile the following VB.NET code:

```vb
Public Class MyClassName
    Private MyStr As String

    Public Function MyFunction() As String
        MyStr = "Hello"
        MyStr# = " World"
        Return MyStr
    End Function
End Class
```

**Expected Behavior**:
The code fails to compile with a compile-time error, *"[Type character][1] '#' does not match declared data type 'String'"* (like it correctly was in e.g. VB.NET 2008).

**Actual Behavior**:
The code compiles and assigns " World" to `MyStr`, ignoring the use of the invalid type character `#`.

Additional reference:
https://stackoverflow.com/q/62105543/11683

[1]: https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/data-types/type-characters

**Comment by jrmoreno1** [comment]
Note that the type character DOES work if applied to a local variable, and you get the expected behavior (BC30277 Type character '#' does not match declared data type 'String'.)

**Comment by AlekseyTs** [comment]
Indeed, native compiler reports
```
(6) : error BC30277: Type character '#' does not match declared data type 'String'.

        MyStr# = " World"
        ~~~~~~
```

**Comment by AlekseyTs** [comment]
For some reason Binder.BindSymbolAccess verfies type character only for MemberAccessExpressionSyntax:
```
                    Dim asMemberAccess = TryCast(node, MemberAccessExpressionSyntax)
                    If asMemberAccess IsNot Nothing AndAlso Not fieldAccessType.IsErrorType() Then
                        VerifyTypeCharacterConsistency(asMemberAccess.Name, fieldAccessType.GetEnumUnderlyingTypeOrSelf, diagnostics)
                    End If
```

This can be observed with the following unit-test:
```
        <Fact()>
        <WorkItem(44724, "https://github.com/dotnet/roslyn/issues/44724")>
        Public Sub Issue44724()
            Dim compilation = CompilationUtils.CreateCompilation(
<compilation>
    <file name="a.vb">
Public Class MyClassName
    Private MyStr As String

    Public Function MyFunction() As String
        MyStr = "Hello"
        MyStr# = " World"
        Me.MyStr# = " World 2"
        Return MyStr
    End Function
End Class
    </file>
</compilation>)
            CompilationUtils.AssertTheseDiagnostics(compilation,
<expected>
BC30277: Type character '#' does not match declared data type 'String'.
        Me.MyStr# = " World 2"
           ~~~~~~
</expected>)
        End Sub
```

Ass far as I can tell, the code was this way since the beginning of Roslyn.

**Comment by cston** [comment]
Unfortunately fixing this bug would be a breaking change.

**Comment by jrmoreno1** [comment]
It was a breaking change when it switched from a compiler error to not, not sure how it is a breaking change to fix a bug so that the code meets the spec.  It won’t compile, but it’s not supposed to.

**Comment by cston** [comment]
> not sure how it is a breaking change to fix a bug so that the code meets the spec.

Visual Basic programs written recently (and compiled with the Roslyn compiler) may be using type characters incorrectly but without errors. If we fix this bug, we'll start reporting errors for those programs.

**Comment by slepmog** [comment]
> Visual Basic programs written recently (and compiled with the Roslyn compiler) may be using type characters incorrectly but without errors. If we fix this bug, we'll start reporting errors for those programs.

Isn't that the correct behaviour? I'm pretty certain that bugs in the C++ compiler are routinely fixed, and when it causes broken code to stop compiling, it is seen as a most useful outcome.

**Comment by CyrusNajmabadi** [comment]
> Isn't that the correct behaviour?

Correct behavior is secondary to back compat.  We could, however, consider re-enabling this in warning wave.  @cston would this fit into that category?
